pkg/llmvalidate/llmvalidate.go

```diff
@@ -0,0 +1,307 @@
+
+* question: The original question
+* answer: The answer
+* files: An array of related files if applicable.
+* short_answer: Yes/No/NA
+* code_snippet: The code snippet relevant to the question. Empty if not applicable
+        `,
+			),
+		},
+	}
+
+	formattedQuestions := ""
+	for _, question := range questions {
+		formattedQuestions += fmt.Sprintf("- %s\n", question)
+	}
+
+	mainPrompt := fmt.Sprintf(`
+The files in the repository are:
+### START OF FILES ###
+
+%s
+
+### END OF FILES ###
+
+Answer the following questions in the context of the code above. be brief in your answers.
+
+%s
+
+`, strings.Join(codePrompt, "\n"), formattedQuestions)
+
+	modelResponse, err := model.GenerateContent(c.ctx, genai.Text(mainPrompt))
+	if err != nil {
+		return nil, err
+	}
+
+	content := getTextContentFromModelContentResponse(modelResponse)
+
+	//unmarshall content into []LLMAnswer
+	var answers []LLMAnswer
+	err = json.Unmarshal([]byte(content), &answers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal content: %v", err)
+	}
+
+	return answers, nil
+
+}
+
+func getTextContentFromModelContentResponse(modelResponse *genai.GenerateContentResponse) string {
+	if len(modelResponse.Candidates) == 0 {
+		return ""
+	}
+	content := modelResponse.Candidates[0].Content
+	finalContent := ""
+	for _, part := range content.Parts {
+		finalContent += fmt.Sprint(part)
+	}
+	return finalContent
+}
+
+func getPromptContentForCode(codePath string) ([]string, error) {
+	var prompts []string
+
+	err := filepath.Walk(codePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		relFilePath, err := filepath.Rel(codePath, path)
+		if err != nil {
+			return err
+		}
+
+		extension := filepath.Ext(relFilePath)
+		if !isAllowedExtension(extension) {
+			return nil
+		}
+
+		if isIgnoredFile(relFilePath) {
+			return nil
+		}
+
+		prompt := getPromptContentForFile(codePath, relFilePath)
+		if prompt != "" {
+			prompts = append(prompts, prompt)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return prompts, nil
+}
+
+func isAllowedExtension(extension string) bool {
+	for _, ext := range allowExtensions {
+		if ext == extension {
+			return true
+		}
+	}
+	return false
+}
+
+func isIgnoredFile(file string) bool {
+	flags := fnmatch.FNM_PERIOD | fnmatch.FNM_NOESCAPE
+	for _, pattern := range ignoreList {
+		if fnmatch.Match(pattern, file, flags) {
+			return true
+		}
+	}
+	return false
+}
+
+func getPromptContentForFile(codePath, relFile string) string {
+	content, err := readFileContent(path.Join(codePath, relFile))
+	if err != nil {
+		logme.DebugFln("Error reading file %s: %v", relFile, err)
+		// we are ignoring this error because this might be a non-text file
+		return ""
+	}
+
+	if !utf8.ValidString(content) {
+		return ""
+	}
+
+	if isMinifiedJsFile(content) {
+		return ""
+	}
+
+	if len(content) == 0 {
+		return ""
+	}
```

Comment:
Maybe we also have to add a max size limit, so if the file is too big it doesn't get sent to the LLM? We would then have to bubble-up this kind of warning to the output even if it's not sent to the LLM, so it may be a bit complicated. I would leave it for a follow-up PR eventually, but interested in hearing if you think this can be a concern

Reply:
THe great thing about gemini is the huge context size. currently above 2M. I am sure we'll eventually hit the limit with some incredibly big plugin but it will work for most cases.

in those cases the LLM will fail with a bad request and after some retries the validator will give up,